langchain/src/tools/dataforseo_api_search.ts

```diff
@@ -0,0 +1,217 @@
+      throw new Error("api_login or api_password is not provided");
+    }
+
+    const credentials = Buffer.from(
+      `${this.apiLogin  }:${  this.apiPassword}`,
+      "utf-8"
+    ).toString("base64");
+    const headers = {
+      Authorization: `Basic ${credentials}`,
+      "Content-Type": "application/json",
+    };
+    const obj: any = { keyword: encodeURIComponent(keyword) };
+    const mergedParams = { ...obj, ...this.params };
+    const data = [mergedParams];
+
+    return {
+      url: `https://api.dataforseo.com/v3/serp/${mergedParams.se_name}/${mergedParams.se_type}/live/advanced`,
+      headers,
+      data,
+    };
+  }
+
+  private async getResponseJson(url: string): Promise<any> {
+    const requestDetails = this.prepareRequest(url);
+    const response = await fetch(requestDetails.url, {
+      method: "POST",
+      headers: requestDetails.headers,
+      body: JSON.stringify(requestDetails.data),
+    });
+
+    if (!response.ok) {
+      throw new Error(
+        `Got ${response.status} error from DataForSEO: ${response.statusText}`
+      );
+    }
+
+    const result = await response.json();
+    return this.checkResponse(result);
+  }
+
+  private checkResponse(response: any): any {
+    if (response.status_code !== 20000) {
+      throw new Error(
+        `Got error from DataForSEO SERP API: ${response.status_message}`
+      );
+    }
+    return response;
+  }
+
+  private filterResults(
+    res: any,
+    types: Array<string> | undefined
+  ): Array<any> {
+    const output: Array<any> = [];
+    for (const task of res.tasks || []) {
+      for (const result of task.result || []) {
+        for (const item of result.items || []) {
+          if (
+            types === undefined ||
+            types.length === 0 ||
+            types.includes(item.type)
+          ) {
+            this.cleanupUnnecessaryItems(item);
+            if (Object.keys(item).length !== 0) {
+              output.push(item);
+            }
+          }
+          if (this.top_count !== undefined && output.length >= this.top_count) {
+            break;
+          }
+        }
+      }
+    }
+    return output;
+  }
+
+  /* eslint-disable no-param-reassign */
+  private cleanupUnnecessaryItems(d: any) {
+    if (this.json_result_fields !== undefined) {
+      for (const key in d) {
+        if (typeof d[key] === "object" && d[key] !== null) {
+          this.cleanupUnnecessaryItems(d[key]);
+          if (Object.keys(d[key]).length === 0) {
+            delete d[key];
+          }
+        } else if (!this.json_result_fields.includes(key)) {
+          delete d[key];
+        }
+      }
+    }
+
+    ["xpath", "position", "rectangle"].forEach((key) => delete d[key]);
+    for (const key in d) {
+      if (typeof d[key] === "object" && d[key] !== null) {
+        this.cleanupUnnecessaryItems(d[key]);
+      }
+    }
+  }
+
+  private processResponse(res: any): string {
+    let toret = "No good search result found";
```

Comment:
I assume this is shorthand for `toReturn` - can we use `returnValue` instead or something more explicit?